File: plugins/logistics/backend/dto/reconciliation.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class ReconciliationCountItemRequest(BaseModel):
    product_id: str
    counted_quantity: float


class ReconciliationCountRequest(BaseModel):
    notes: str | None = Field(default=None, max_length=500)
    items: list[ReconciliationCountItemRequest]


class CloseSessionRequest(BaseModel):
    notes: str | None = Field(default=None, max_length=500)


class ReconciliationLineRead(BaseModel):
    product_id: str
    product_name: str
    expected_quantity: float
    counted_quantity: float | None = None
    difference_quantity: float | None = None


class InventoryDiscrepancyRead(BaseModel):
    id: str
    product_id: str
    product_name: str
    expected_quantity: float
    counted_quantity: float
    difference_quantity: float
    status: str
    resolution_notes: str | None = None


class SessionReconciliationRead(BaseModel):
    id: str | None = None
    session_id: str
    status: str
    counted_by: str | None = None
    counted_at: datetime | None = None
    notes: str | None = None
    can_close: bool = False
    lines: list[ReconciliationLineRead] = Field(default_factory=list)
    discrepancies: list[InventoryDiscrepancyRead] = Field(default_factory=list)

File: plugins/logistics/frontend/api/reconciliation.ts
import { API_PREFIX } from "./_shared";
import { apiRequest } from "@systutor/shell/api/client";

export type ReconciliationLine = {
  product_id: string;
  product_name: string;
  expected_quantity: number;
  counted_quantity: number | null;
  difference_quantity: number | null;
};

export type InventoryDiscrepancy = {
  id: string;
  product_id: string;
  product_name: string;
  expected_quantity: number;
  counted_quantity: number;
  difference_quantity: number;
  status: string;
  resolution_notes: string | null;
};

export type SessionReconciliation = {
  id: string | null;
  session_id: string;
  status: string;
  counted_by: string | null;
  counted_at: string | null;
  notes: string | null;
  can_close: boolean;
  lines: ReconciliationLine[];
  discrepancies: InventoryDiscrepancy[];
};

export type ReconciliationCountPayload = {
  notes?: string | null;
  items: Array<{
    product_id: string;
    counted_quantity: number;
  }>;
};

export type CloseSessionPayload = {
  notes?: string | null;
};

export function getSessionReconciliation(sessionId: string) {
  return apiRequest<SessionReconciliation>(`${API_PREFIX}/vehicle-sessions/${sessionId}/reconciliation`);
}

export function countSessionReconciliation(sessionId: string, payload: ReconciliationCountPayload) {
  return apiRequest<SessionReconciliation>(`${API_PREFIX}/vehicle-sessions/${sessionId}/reconciliation/count`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function closeVehicleSession(sessionId: string, payload: CloseSessionPayload = {}) {
  return apiRequest<{ session_id: string; status: string }>(`${API_PREFIX}/vehicle-sessions/${sessionId}/close`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
